python/ray/data/read_api.py

```diff
@@ -376,28 +385,51 @@ def read_datasource(
             )
         )
 
-    if read_tasks and len(read_tasks) < min_safe_parallelism * 0.7:
-        perc = 1 + round((min_safe_parallelism - len(read_tasks)) / len(read_tasks), 1)
-        logger.warning(
-            f"{WARN_PREFIX} The blocks of this dataset are estimated to be {perc}x "
-            "larger than the target block size "
-            f"of {int(ctx.target_max_block_size / 1024 / 1024)} MiB. This may lead to "
-            "out-of-memory errors during processing. Consider reducing the size of "
-            "input files or using `.repartition(n)` to increase the number of "
-            "dataset blocks."
-        )
-    elif len(read_tasks) < requested_parallelism and (
-        len(read_tasks) < ray.available_resources().get("CPU", 1) // 2
-    ):
-        logger.warning(
-            f"{WARN_PREFIX} The number of blocks in this dataset "
-            f"({len(read_tasks)}) "
-            f"limits its parallelism to {len(read_tasks)} concurrent tasks. "
-            "This is much less than the number "
-            "of available CPU slots in the cluster. Use `.repartition(n)` to "
-            "increase the number of "
-            "dataset blocks."
-        )
+    #    if read_tasks and len(read_tasks) < min_safe_parallelism * 0.7:
+    #        perc = 1 + round((min_safe_parallelism - len(read_tasks)) / len(read_tasks), 1)
+    #        logger.warning(
+    #            f"{WARN_PREFIX} The blocks of this dataset are estimated to be {perc}x "
+    #            "larger than the target block size "
+    #            f"of {int(ctx.target_max_block_size / 1024 / 1024)} MiB. This may lead to "
+    #            "out-of-memory errors during processing. Consider reducing the size of "
+    #            "input files or using `.repartition(n)` to increase the number of "
+    #            "dataset blocks."
+    #        )
+    #    elif len(read_tasks) < requested_parallelism and (
+    #        len(read_tasks) < ray.available_resources().get("CPU", 1) // 2
+    #    ):
+    #        logger.warning(
+    #            f"{WARN_PREFIX} The number of blocks in this dataset "
+    #            f"({len(read_tasks)}) "
+    #            f"limits its parallelism to {len(read_tasks)} concurrent tasks. "
+    #            "This is much less than the number "
+    #            "of available CPU slots in the cluster. Use `.repartition(n)` to "
+    #            "increase the number of "
+    #            "dataset blocks."
+    #        )
+
+    # TODO update the warnings above
+    if len(read_tasks) < requested_parallelism:
```

Comment:
`len(read_tasks) < requested_parallelism` is one case. another case that needs splitting blocks is `len(read_tasks) < downstream parallelism (e.g., actor pool size)`.  I think we'll need an optimizer rule for this anyway. we can also move the `_autodetect_parallelism` logic to this rule.

Reply:
Yeah, agreed we should move autoparallelism to a rule (probably after removing all the legacy code).

I'm not sure there's much value in checking `len(read_tasks) < downstream parallelism` though, since we already check `len(read_tasks) < ncpu * 2`, and ncpu is an upper bound on downstream parallelism.

Reply:
I think this case is not covered by the current approach: `ray.data.read_parquet(..., parallelism=10).map(..., compute=ActorPoolStrategy(size=100))`. 
I think this usage make sense because if the read is cheap while the map is expensive, we may want to set a larger parallelism for the map.

Reply:
Right, but that's the user's fault. We heavily discourage setting parallelism, except in extreme cases.